pkg/ingress/gateway.go

```diff
@@ -17,66 +16,10 @@ import (
 	"github.com/openservicemesh/osm/pkg/certificate"
 )
 
-// provisionIngressGatewayCert does the following:
-// 1. If an ingress gateway certificate spec is specified in the MeshConfig resource, issues a certificate
-//    for it and stores it in the referenced secret.
-// 2. Starts a goroutine to watch for changes to the MeshConfig resource and certificate rotation, and
-//    updates/removes the certificate and secret as necessary.
-func (c *client) provisionIngressGatewayCert(stop <-chan struct{}) error {
-	defaultCertSpec := c.cfg.GetMeshConfig().Spec.Certificate.IngressGateway
-	if defaultCertSpec != nil {
-		// Issue a certificate for the default certificate spec
-		if err := c.createAndStoreGatewayCert(*defaultCertSpec); err != nil {
-			return errors.Wrap(err, "Error provisioning default ingress gateway cert")
-		}
-	}
-
-	// Initialize a watcher to watch for CREATE/UPDATE/DELETE on the ingress gateway cert spec
-	go c.handleCertificateChange(defaultCertSpec, stop)
-
-	return nil
-}
-
-// createAndStoreGatewayCert creates a certificate for the given certificate spec and stores
-// it in the referenced k8s secret if the spec is valid.
-func (c *client) createAndStoreGatewayCert(spec configv1alpha2.IngressGatewayCertSpec) error {
-	if len(spec.SubjectAltNames) == 0 {
-		return errors.New("Ingress gateway certificate spec must specify at least 1 SAN")
-	}
-
-	// Validate the validity duration
-	if _, err := time.ParseDuration(spec.ValidityDuration); err != nil {
-		return errors.Wrapf(err, "Invalid cert duration '%s' specified", spec.ValidityDuration)
-	}
-
-	// Validate the secret ref
-	if spec.Secret.Name == "" || spec.Secret.Namespace == "" {
-		return errors.Errorf("Ingress gateway cert secret's name and namespace cannot be nil, got %s/%s", spec.Secret.Namespace, spec.Secret.Name)
-	}
-
-	// Issue a certificate
-	// OSM only support configuring a single SAN per cert, so pick the first one
-	certCN := spec.SubjectAltNames[0]
-
-	// A certificate for this CN may be cached already. Delete it before issuing a new certificate.
-	c.certProvider.ReleaseCertificate(certCN)
-	issuedCert, err := c.certProvider.IssueCertificate(certCN, certificate.IngressGateway, certificate.FullCNProvided())
-	if err != nil {
-		return errors.Wrapf(err, "Error issuing a certificate for ingress gateway")
-	}
-
-	// Store the certificate in the referenced secret
-	if err := c.storeCertInSecret(issuedCert, spec.Secret); err != nil {
-		return errors.Wrapf(err, "Error storing ingress gateway cert in secret %s/%s", spec.Secret.Namespace, spec.Secret.Name)
-	}
-
-	return nil
-}
-
 // storeCertInSecret stores the certificate in the specified k8s TLS secret
 func (c *client) storeCertInSecret(cert *certificate.Certificate, secret corev1.SecretReference) error {
 	secretData := map[string][]byte{
-		"ca.crt":  cert.GetIssuingCA(),
+		"ca.crt":  cert.GetTrustedCAs(),
```

Comment:
This is specifically for the cert the ingress gateway will present to the upstream service; why does the ingress gateway need the whole cert chain?

Reply:
If the ingress gateway is doing mTLS, wouldn't it need both CA's during a rotation in case it is misaligned from the upstream svc?

Reply:
Ah you're right; the ingress gateway only sends, but it does need to validate the certificate of the receiver 👍🏾